core/src/main/java/ai/timefold/solver/core/impl/domain/variable/listener/support/VariableListenerSupport.java

```diff
@@ -240,61 +261,83 @@ public void beforeListVariableChanged(ListVariableDescriptor<Solution_> variable
-                // Evaluate all elements inside the range
-                evaluateFromIndex(values, event.fromIndex(), event.toIndex(), true, cascadingUpdateShadowVariableDescriptor);
-                // Evaluate later elements, but stops when there is no change
-                evaluateFromIndex(values, event.toIndex(), values.size(), false, cascadingUpdateShadowVariableDescriptor);
-            }
+            cascadeListVariableChangedNotifications(cascadingUpdateShadowVariableDescriptor);
+            // When the unassigned element has no inverse entity,
+            // it indicates that it is not reverting to a previous entity.
+            // In this case, we need to invoke the cascading logic,
+            // or its related shadow variables will remain unchanged.
+            cascadeUnassignedValues(cascadingUpdateShadowVariableDescriptor);
+        }
+        listVariableChangedNotificationList.clear();
+        unassignedValueWithEmptyInverseEntitySet.clear();
+    }
+
+    private void cascadeListVariableChangedNotifications(
+            CascadingUpdateShadowVariableDescriptor<Solution_> cascadingUpdateShadowVariableDescriptor) {
+        for (var notification : listVariableChangedNotificationList) {
+            cascadeListVariableValueUpdates(
+                    listVariableDescriptor.getValue(notification.getEntity()),
+                    notification.getFromIndex(), notification.getToIndex(),
+                    cascadingUpdateShadowVariableDescriptor);
         }
     }
 
-    private void evaluateFromIndex(List<Object> values, int fromIndex, int toIndex, boolean forceUpdate,
+    private void cascadeListVariableValueUpdates(List<Object> values, int fromIndex, int toIndex,
             CascadingUpdateShadowVariableDescriptor<Solution_> cascadingUpdateShadowVariableDescriptor) {
-        var lastUpdated = fromIndex;
-        while (lastUpdated < toIndex) {
-            if (!cascadingUpdateShadowVariableDescriptor.update(scoreDirector, values.get(lastUpdated))
-                    && !forceUpdate) {
+        for (int currentIndex = fromIndex; currentIndex < values.size(); currentIndex++) {
+            var value = values.get(currentIndex);
+            // The value is present in the unassigned values,
+            // but the cascade logic is triggered by a list event.
+            // So, we can remove it from the unassigned list
+            // since the entity will be reverted to a previous entity.
+            unassignedValueWithEmptyInverseEntitySet.remove(value);
+            // Force updates within the range.
+            // Outside the range, only update while the values keep changing.
+            var forceUpdate = currentIndex < toIndex;
+            if (!cascadingUpdateShadowVariableDescriptor.update(scoreDirector, value) && !forceUpdate) {
                 break;
             }
-            lastUpdated++;
+        }
+    }
+
+    private void cascadeUnassignedValues(
```

Comment:
I only split this into a separate method so that I can see its impact in the profiler individually. This is not actually a perf optimization, just helps with profiling.